[[ Bug 22143 ]] Fix interactive tutorial on Windows

livecodepanos opened the pull request:
This patch forces a redraw when showing an invisible stack that has a windowshape. Previously, to fix [bug 21805](https://github.com/livecode/livecode/pull/6860/files) we used `loadwindowshape()` for this, but this caused a problem with the interactive tutorial on Windows ([bug 22143](https://quality.livecode.com/show_bug.cgi?id=22143)). This patch ensures both bugs are addressed.

runrevmark reviewed (commented):
I don't think this fix can be correct. The existing logic was:
```
if window has windowShape property set then
   destroy existing window shape (as it is about to be reset)
   if image found then
       create new mask from currently set image
       if mask was created then
           update window's mask with it
       end if
   end if
end if
```
So this will work as long as (1) the image is found and (2) the image can have its mask extracted (which I think only fails due to memory alloc failing - although not 100% sure on that). All images can generate a mask (if the image has no alpha channel its just a solid rectangle).

The existing code ensures that the windowshape is removed regardless of whether the new setting's image exists, or fails to generate a mask - the problem is that doing that is not causing the window to be refreshed without a window shape (I would guess).

At what point following that line is the method failin
  - is t_image nullptr?
  - is t_new_mask nullptr?
  - What is the windowshapeid of the window at this point?
  - What does the image it is set to look like?

Also, where is `loadwindowshape()` being called from in this instance?

Note: Windows and macOS are somewhat different in how windowshapes work - on Windows a windowshape can be sharp (meaning 1-bit) which is then done using a clipping region *or* it can be soft (alpha) which is done using a layered window - these two methods use slightly different code lower down. IIRC an image with either alpha 0 or 255 in every pixel is considered sharp so which the mask in this case is, might help get to the bottom of the problem.

livecodepanos commented:
What the interactive tutorial does is:

1. sets the windowshape of the tutorial stack to a number (non-zero)
2. shows the stack (the stack has instructions on how to do each step of the tutorial)
3. hides the stack when you do what the instructions say
4. shows the stack (which now has new instructions) and so on  

The problem is caused in step 4, i.e. when the stack is set to visible after being hidden.

In the `MCStack::SetVisible()` method, the `loadwindowshape()` is called (see https://github.com/livecode/livecode/pull/6860/files).

At this point:

`t_image` is _not_ nullptr
`t_new_mask` is nullptr 

Let me check what is the windowshapeid of the window at this point

runrevmark commented:
Okay so that suggests the image exists - so see if you can find out where in the code the mask is being created it fails (thus making it return nullptr); and also see if you can find out what image is being referred to (by the windowshapeid).

livecodepanos commented:
Ok, so the `windowshapeid` is `225105` when this happens. The code in the mask being created fails when `lockbitmap()` is called in `MCImage::makewindowshape()` in w32image.cpp. The resulting `t_bitmap` has `t_bitmap ->  has_transparency = false` as well as  `t_bitmap ->  has_alpha = false`, so `MCImage::makewindowshape()` returns `nil`.

runrevmark commented:
@livecodepanos So it occurs to me that the problem here is probably that a windowshape is *not* being generated for that image.

Can you get the 'text' of the image which is failing, just so we can easily test what the image is?

I suspect if we make MCImage::makewindowshape (w32 version) create a mask for the image things might work...

Indeed, looking at BitmapToRegion I *think* it will work whether the image has a mask or not (all images in LC are unpacked as ARGB) so perhaps just try removing `if (t_has_mask)` so the BitmapToRegion call is used when the image doesn't have an alpha channel?

runrevmark commented on engine/src/exec-interface-stack.cpp:
```diff
@@ -485,9 +485,9 @@ void MCStack::SetVisible(MCExecContext& ctxt, uint32_t part, bool setting)
-        if (setting)
-            loadwindowshape();
+        // On Windows force a redraw, otherwise a stack with a windowshape does not show
+        if (setting && windowshapeid)
```
Make the fact that `0` windowshapeid means 'no window shape' explicit by doing:
```
if (setting && (windowshapeid != 0))
```

runrevmark reviewed (changes requested):
Please make the commit title start with `[[ Bug 22143 ]]`

livecodepanos committed:
[[ Bug 22143 ]] Redraw window with windowShape after being reshown

This patch forces a redraw when showing an invisible stack
that has a windowshape. Previously, to fix bug 21805 we used
`loadwindowshape()` for this, but this caused a problem with
the interactive tutorial on Windows (bug 22143). This patch
ensures both bugs are addressed.

runrevmark reviewed (changes requested):
So from a more general point of view, this fix isn't about fixing the interactive tutorial on windows - that's just a specific user-visible outcome.

The release note is fine as it is but commit messages/titles should indicate what was actually done; rather than (the positive) side-effects. So:
```
[[ Bug 22143 ]] Redraw window with windowShape after being reshown
```

(Sorry - this didn't occur to me before!)

runrevmark commented:
@livecode-vulcan review ok 4881b70

livecode-vulcan commented:
💙 review by @runrevmark ok 4881b7042a5914d064c0140dfa5d2b2e6a1324ea

livecode-vulcan commented:
😎 test success 4881b7042a5914d064c0140dfa5d2b2e6a1324ea

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/857)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/266)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/266)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/266)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/266)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2383)
* try-community-universal-ios-iphoneos13.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos13.5/builds/34)
* try-community-universal-ios-iphonesimulator13.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator13.5/builds/34)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2411)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2399)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2398)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2433)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2373)
